vowpalwabbit/plt.cc

```diff
@@ -0,0 +1,382 @@
+    fp = 0;
+    fn = 0;
+    ec_count = 0;
+    true_count = 0;
+  }
+
+  ~plt()
+  {
+    nodes_time.delete_v();
+    node_preds.delete_v();
+    tp_at.delete_v();
+  }
+};
+
+inline void learn_node(plt& p, uint32_t n, single_learner& base, example& ec)
+{
+  if (!p.all->weights.adaptive)
+  {
+    p.all->sd->t = p.nodes_time[n];
+    p.nodes_time[n] += ec.weight;
+  }
+  base.learn(ec, n);
+}
+
+void learn(plt& p, single_learner& base, example& ec)
+{
+  MULTILABEL::labels multilabels = std::move(ec.l.multilabels);
+  MULTILABEL::labels preds = std::move(ec.pred.multilabels);
+
+  double t = p.all->sd->t;
+  double weighted_holdout_examples = p.all->sd->weighted_holdout_examples;
+  p.all->sd->weighted_holdout_examples = 0;
+
+  std::unordered_set<uint32_t> n_positive;  // positive nodes
+  std::unordered_set<uint32_t> n_negative;  // negative nodes
+
+  if (multilabels.label_v.size() > 0)
+  {
+    for (auto label : multilabels.label_v)
+    {
+      uint32_t tn = label + p.ti;
+      if (tn < p.t)
+      {
+        n_positive.insert(tn);
+        while (tn > 0)
+        {
+          tn = floor(static_cast<float>(tn - 1) / p.kary);
+          n_positive.insert(tn);
+        }
+      }
+    }
+    if (ec.l.multilabels.label_v.last() >= p.k)
+      std::cout << "label " << ec.l.multilabels.label_v.last() << " is not in {0," << p.k - 1
+                << "} This won't work right." << std::endl;
+
+    for (auto& n : n_positive)
+    {
+      if (n < p.ti)
+      {
+        for (uint32_t i = 1; i <= p.kary; ++i)
+        {
+          uint32_t n_child = p.kary * n + i;
+          if (n_child < p.t && n_positive.find(n_child) == n_positive.end())
+            n_negative.insert(n_child);
+        }
+      }
+    }
+  }
+  else
+    n_negative.insert(0);
+
+  ec.l.simple = {1.f, 1.f, 0.f};
+  for (auto& n : n_positive) learn_node(p, n, base, ec);
+
+  ec.l.simple.label = -1.f;
+  for (auto& n : n_negative) learn_node(p, n, base, ec);
+
+  p.all->sd->t = t;
+  p.all->sd->weighted_holdout_examples = weighted_holdout_examples;
+
+  ec.pred.multilabels = std::move(preds);
+  ec.l.multilabels = std::move(multilabels);
+}
+
+inline float predict_node(uint32_t n, single_learner& base, example& ec)
+{
+  ec.l.simple = {FLT_MAX, 1.f, 0.f};
+  base.predict(ec, n);
+  return 1.0f / (1.0f + exp(-ec.partial_prediction));
+}
+
+template <bool threshold>
+void predict(plt& p, single_learner& base, example& ec)
+{
+  MULTILABEL::labels multilabels = std::move(ec.l.multilabels);
+  MULTILABEL::labels preds = std::move(ec.pred.multilabels);
+  preds.label_v.clear();
+
+  // split labels into true and skip (those > max. label num)
+  std::unordered_set<uint32_t> true_labels;
+  std::unordered_set<uint32_t> skip_labels;
+  for (auto label : multilabels.label_v)
+  {
+    if (label < p.k)
+      true_labels.insert(label);
+    else
+      skip_labels.insert(label - p.k);
+  }
+
+  // prediction with threshold
+  if (threshold)
+  {
+    std::queue<node> node_queue;
```

Comment:
>queue [](start = 9, length = 5)

This will also cause allocations so would need similar treatment.  i.e. create a container in plt that reaches steady state size and does not cause allocations after that.

Reply:
Change it to a simple vector and member of `plt` structure.